src/snowcli/cli/common/cli_global_context.py

```diff
@@ -0,0 +1,146 @@
+from dataclasses import dataclass
+from typing import Optional, Union, Callable, Any
+
+from snowflake.connector import SnowflakeConnection
+
+from snowcli.output.formats import OutputFormat
+from snowcli.snow_connector import connect_to_snowflake
+
+
+@dataclass
+class ConnectionDetails:
+    connection_name: Optional[str] = None
+    account: Optional[str] = None
+    database: Optional[str] = None
+    role: Optional[str] = None
+    schema: Optional[str] = None
+    user: Optional[str] = None
+    password: Optional[str] = None
+    authenticator: Optional[str] = None
+    private_key_path: Optional[str] = None
+    warehouse: Optional[str] = None
+    temporary_connection: bool = False
+
+    def _resolve_connection_params(self):
+        from snowcli.cli.common.decorators import GLOBAL_CONNECTION_OPTIONS
+
+        params = {}
+        for option in GLOBAL_CONNECTION_OPTIONS:
+            override = option.name
+            if override == "connection" or override == "temporary_connection":
+                continue
+            override_value = getattr(self, override)
+            if override_value is not None:
+                params[override] = override_value
+        return params
+
+    def build_connection(self):
+        return connect_to_snowflake(
+            temporary_connection=self.temporary_connection,
+            connection_name=self.connection_name,
+            **self._resolve_connection_params()
+        )
+
+
+DEFAULT_ENABLE_TRACEBACKS = True
+DEFAULT_OUTPUT_FORMAT = OutputFormat.TABLE
+DEFAULT_VERBOSE = False
+DEFAULT_EXPERIMENTAL = False
+
+
+class _GlobalContextManager:
+    _connection_details = ConnectionDetails()
+    _cached_connection: Optional[SnowflakeConnection] = None
+
+    enable_tracebacks = DEFAULT_ENABLE_TRACEBACKS
+    output_format = DEFAULT_OUTPUT_FORMAT
+    verbose = DEFAULT_VERBOSE
+    experimental = DEFAULT_EXPERIMENTAL
+
+    def set_default_values(self):
+        self.connection_details = ConnectionDetails()
+        self.enable_tracebacks = DEFAULT_ENABLE_TRACEBACKS
+        self.output_format = DEFAULT_OUTPUT_FORMAT
+        self.verbose = DEFAULT_VERBOSE
+        self.experimental = DEFAULT_EXPERIMENTAL
+
+    @property
+    def connection_details(self) -> ConnectionDetails:
+        return self._connection_details
+
+    @connection_details.setter
+    def connection_details(self, connection_details: ConnectionDetails):
+        self._connection_details = connection_details
+        self._cached_connection = None
+
+    @property
+    def cacheable_connection(self) -> SnowflakeConnection:
```

Comment:
Do we plan to have other connection? Maybe we can drop the `cachable` prefix?

Reply:
Done